Fix iscsi logout issues

jsafrane opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Our iSCSI e2e tests sometimes report that a file that has been just written to an iSCSI volume is not available there. e2e systems are chronically overloaded and I noticed two issues there:

* iscsiadm sometimes reports `iscsiadm: Could not execute operation on all records: encountered iSCSI database failure`. We should check all iscsiadm exit codes and re-try logging out from a target. 
  * This change can be tricky, we ignored iscsiadm errors previously and now we may enter endless loop if iscsiadm fails constantly. kubelet should cope with that IMO. I added extra checks that make sure that logging off from already removed target succeeds.

* Looking at Red Hat docs, it suggests flushing devices before removing them:
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/storage_administration_guide/removing_devices. I added `blockdev --flushbufs <device>` to make sure we don't loose data, especially when using raw block devices.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Failed iscsi logout is now re-tried periodically.
```

jsafrane commented:
/test pull-kubernetes-e2e-gce-iscsi

jsafrane commented:
/test pull-kubernetes-e2e-gce-iscsi

jsafrane commented:
/test pull-kubernetes-e2e-gce-iscsi

jsafrane commented:
/test pull-kubernetes-e2e-gce-iscsi

jsafrane commented:
Rebased and removed testing patches. Now it's ready for review

@j-griffith @bswartz @humblec @bertinatto, PTAL

jsafrane commented:
/test pull-kubernetes-node-e2e-containerd/test 

(iscsi tests are broken because of #78940)

bertinatto commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -740,23 +751,29 @@ func (util *ISCSIUtil) detachISCSIDisk(exec mount.Exec, portals []string, iqn, i
+		err = ignoreExitCodes(err, exit_ISCSI_ERR_NO_OBJS_FOUND, exit_ISCSI_ERR_SESS_NOT_FOUND)
 		if err != nil {
 			klog.Errorf("iscsi: failed to delete node record Error: %s", string(out))
+			return err
```
The changes LGTM, specially the device flushing; however, IDK iSCSI well enough to be sure these are the only exit codes we could/should ignore.

For example, looking through the exit codes it seems we should always etry on `ISCSI_ERR_AGAIN` and `ISCSI_ERR_BUSY`, but not for `ISCSI_ERR_ACCESS` and `ISCSI_ERR_INVAL`. To make it difficult, these codes are tied to a specific set of commands.

bertinatto replied:
Anyway, I think it's OK to retry regardless

jsafrane replied:
I don't see ISCSI_ERR_AGAIN and ISCSI_ERR_BUSY in my man page (iscsiadm version 6.2.0.876-1)

jsafrane replied:
I saw `ISCSI_ERR_NO_OBJS_FOUND` for deleting already deleted session and I think I saw `ISCSI_ERR_SESS_NOT_FOUND` in some log.

bertinatto replied:
I was looking at the [iscsi_err.h](https://github.com/open-iscsi/open-iscsi/blob/7d121d12ad6ba7783308c25ffd338a9fa0cc402b/include/iscsi_err.h#L7) file, so maybe these exit codes aren't returned by iscsiadm.

j-griffith replied:
There are a bunch of different things that can be returned from iscsiadm that may or may not matter.  One way to approach this is ignore any error, and do a verify that things were cleaned up, if it has the error is irrelevant.

jsafrane replied:
> verify that things were cleaned up

I can check that iface and node need to be deleted easily (`-m [node|iface] -o show`).

I am not sure if it's possible to check if `iscsiadm -m node -p 127.0.0.1 -T iqn.2003-01.io.k8s:e2e.volume-8751 --logout` is needed. `iscsiadm -m node -o show` is the same before and after logout. How can I map node / iqn to a session without too much parsing of `iscsiadm -m session -o show -P 2`?

j-griffith replied:
Yeah, I was thinking of using the approach to check for sessions; I may be overthinking this though.

j-griffith reviewed (approved):
I think this is "ok" as is, but if you wanted to be sure about the iscsiadm errors I'd suggest just removing the explicit checks for errors being returned, instead just if there is an error check if thing were cleaned up or not.

bswartz commented:
Flushing the device happens automatically on unmount. Is this bug about raw block iSCSI? Or is there good evidence that unmount by itself isn't good enough?

jsafrane commented:
Unmount should be enough, this is extra precaution for block.

jsafrane committed:
Report error when iscsiadm fails during detach

Kubernetes should retry detaching iSCSI volumes on error. In addition, it
should not report an error when detaching a disk while the disk is already
detached.

jsafrane commented:
/retest

j-griffith commented:
/lgtm

humblec commented:
/retest

jsafrane commented:
/test pull-kubernetes-integration
